iphone/Maps/Core/iCloud/CloudStorageManager.swift

```diff
@@ -144,17 +144,19 @@ private extension CloudStorageManager {
     guard let error else { return }
     settings.setICLoudSynchronizationEnabled(false)
     synchronizationError = error
-    MWMAlertViewController.activeAlert().presentBugReportAlert(withTitle: L("icloud_synchronization_error_alert_title"))
+    DispatchQueue.main.async {
+      MWMAlertViewController.activeAlert().presentBugReportAlert(withTitle: L("icloud_synchronization_error_alert_title"))
```

Comment:
Why is it async but on the main thread? A comment would be helpful.

Reply:
DispatchQueue.main.**sync**  called from the main will deadlock the main thread.

DispatchQueue.main.async is more convenient to add the task to the next runloop cycle.

Reply:
I found the issue with the previous solution that some places can throw an error during the file writing on the BG thread. So the `stopSynchronization` method can be called both from the main and bg threads but it's important to always handle the UI alert on the main.

Reply:
I've made a small refactoring for this case and here is the changes:
1. all the completion handlers that are called after work finishsing on the BG thread will automatically dispatched **synchronously** back to the main (because call disp.main.sync is safe when called from the different queue and do not block the main). So it is always safe to work with the file writing results and process errors or reload the bookmarks using the bookmarks manager. The overa cloud sync work is processed on the main except the file writing/copyng/replacing etc - this tasks are dispatched to the bg thread.
2. all the unnecessary calls of the disp.main.async was removed